libdevice/cmake/modules/SYCLLibdevice.cmake

```diff
@@ -29,152 +20,138 @@ set(compile_opts
-                           -o ${spv_binary_dir}/libsycl-fallback-cmath.spv
-                   MAIN_DEPENDENCY fallback-cmath.cpp
-                   DEPENDS device_math.h device.h clang llvm-spirv
-                   VERBATIM)
-
-add_custom_command(OUTPUT ${obj_binary_dir}/libsycl-fallback-cmath.${lib-suffix}
-                   COMMAND ${clang} -fsycl -c
-                           ${compile_opts} ${sycl_targets_opt}
-                           ${CMAKE_CURRENT_SOURCE_DIR}/fallback-cmath.cpp
-                           -o ${obj_binary_dir}/libsycl-fallback-cmath.${lib-suffix}
-                   MAIN_DEPENDENCY fallback-cmath.cpp
-                   DEPENDS device_math.h device.h clang clang-offload-bundler
-                   VERBATIM)
-
-add_custom_command(OUTPUT ${spv_binary_dir}/libsycl-fallback-cmath-fp64.spv
-                   COMMAND ${clang} -S -fsycl-device-only -fno-sycl-use-bitcode
-                           ${compile_opts}
-                           ${CMAKE_CURRENT_SOURCE_DIR}/fallback-cmath-fp64.cpp
-                           -o ${spv_binary_dir}/libsycl-fallback-cmath-fp64.spv
-                   MAIN_DEPENDENCY fallback-cmath-fp64.cpp
-                   DEPENDS device_math.h device.h clang llvm-spirv
-                   VERBATIM)
-
-add_custom_command(OUTPUT ${obj_binary_dir}/libsycl-fallback-cmath-fp64.${lib-suffix}
-                   COMMAND ${clang} -fsycl -c
-                           ${compile_opts} ${sycl_targets_opt}
-                           ${CMAKE_CURRENT_SOURCE_DIR}/fallback-cmath-fp64.cpp
-                           -o ${obj_binary_dir}/libsycl-fallback-cmath-fp64.${lib-suffix}
-                   MAIN_DEPENDENCY fallback-cmath-fp64.cpp
-                   DEPENDS device_math.h device.h clang clang-offload-bundler
-                   VERBATIM)
+function(BuildSYCLDeviceLib)
+  cmake_parse_arguments(DeviceLib "" "Path;Target;Source" "Depends" ${ARGN})
+  add_custom_command(OUTPUT ${DeviceLib_Path}
+                     COMMAND ${clang} -fsycl -fsycl-device-only -emit-llvm
+                             ${compile_opts} -fsycl-targets=${DeviceLib_Target}
+                             ${CMAKE_CURRENT_SOURCE_DIR}/${DeviceLib_Source}
+                             -o ${DeviceLib_Path} -fno-sycl-device-lib=all
+                     MAIN_DEPENDENCY ${DeviceLib_Source}
+                     DEPENDS ${DeviceLib_Depends}
+                     VERBATIM)
+endfunction()
+
+
+set(devicelib-wrapper-crt-spir64 ${obj_binary_dir}/libsycl-crt-spir64.bc)
+BuildSYCLDeviceLib(Path ${devicelib-wrapper-crt-spir64} Target spir64-unknown-unknown-sycldevice Source ${libcrt_source} Depends wrapper.h device.h spirv_vars.h clang)
+BuildSYCLDeviceLib(Path ${obj_binary_dir}/libsycl-fallback-cassert-spir64.bc Target spir64-unknown-unknown-sycldevice Source fallback-cassert.cpp Depends wrapper.h device.h spirv_vars.h clang)
+
+set(devicelib-wrapper-crt-spir64_x86_64 ${obj_binary_dir}/libsycl-crt-spir64_x86_64.bc)
```

Comment:
We can probably avoid duplication by using [foreach](https://cmake.org/cmake/help/v3.0/command/foreach.html) to loop over all targets.

Reply:
Fixed.